libs/src/services/ethContracts/index.js

```diff
@@ -159,70 +159,80 @@ class EthContracts {
+    if (!this.expectedServiceVersions) {
+      this.expectedServiceVersions = await this.getExpectedServiceVersions()
+    }
+    if (!this.expectedServiceVersions.hasOwnProperty(spType)) {
+      throw new Error(`Invalid service name: ${spType}`)
+    }
 
-            // Compare chain service name
-            if (!this.expectedServiceVersions) {
-              this.expectedServiceVersions = await this.getExpectedServiceVersions()
-            }
-            if (!this.expectedServiceVersions.hasOwnProperty(serviceName)) {
-              throw new Error(`Invalid service name: ${serviceName}`)
-            }
+    let expectedVersion = this.expectedServiceVersions[spType]
 
-            if (serviceName !== spType) {
-              throw new Error(`Invalid service type: ${serviceName}. Expected ${spType}`)
-            }
+    let selectedServiceProvider
+    try {
+      // Querying all versions
+      let foundVersions = new Set()
+      let spVersionToEndpoint = {}
+      await Promise.all(serviceProviders.map(async (sp) => {
+        const {
+          data: { service: serviceName, version: serviceVersion }
+        } = await axios(
+          { url: urlJoin(sp.endpoint, 'version'), method: 'get' }
+        )
+        // Discovery provider specific validation
+        if (spType === 'discovery-provider') {
+          await this.validateDiscoveryProviderHealth(sp.endpoint)
+        }
 
-            if (!semver.valid(serviceVersion)) {
-              throw new Error(`Invalid semver version found - ${serviceVersion}`)
-            }
+        if (serviceName !== spType) {
+          throw new Error(`Invalid service type: ${serviceName}. Expected ${spType}`)
+        }
 
-            let expectedVersion = this.expectedServiceVersions[serviceName]
-            if (expectedVersion !== serviceVersion) {
-              let validSPVersion = this.isValidSPVersion(expectedVersion, serviceVersion)
-              if (!validSPVersion) {
-                throw new Error(`Invalid service version: ${serviceName}. Expected ${expectedVersion}, found ${serviceVersion}`)
-              }
-            }
+        if (!semver.valid(serviceVersion)) {
+          throw new Error(`Invalid semver version found - ${serviceVersion}`)
+        }
 
-            return discprov.endpoint
-          } catch (err) {
-            throw new Error(err)
+        if (expectedVersion !== serviceVersion) {
+          // Confirm this returned value is valid, ignoring patch version in semantic versioning
+          let validSPVersion = this.isValidSPVersion(expectedVersion, serviceVersion)
+          if (!validSPVersion) {
+            throw new Error(`Invalid latest service version: ${serviceName}. Expected ${expectedVersion}, found ${serviceVersion}`)
           }
-        })
-      )
+        }
+
+        foundVersions.add(serviceVersion)
+        // Update mapping of version <-> [endpoint], creating array if needed
+        if (!spVersionToEndpoint.hasOwnProperty(serviceVersion)) {
```

Comment:
is this type of concurrent dictionary modification safe...? if this needs a redis lock or something to avoid any issues, that is an option along with removing the promise.all and just iterating linearly (but that will be slower)

Reply:
this should be okay